rdata/conversion/to_r.py

```diff
@@ -75,269 +141,453 @@ def build_r_object(
+                "levels": data.categories.to_numpy(),
+                "class": "factor",
+            }
+
+        elif isinstance(data, pd.DataFrame):
+            is_object = True
+            r_type = RObjectType.VEC
+            column_names = []
+            r_value = []
+            for column, series in data.items():
+                assert isinstance(column, str)
+                column_names.append(column)
+
+                pd_array = series.array
+                array: pd.Categorical | npt.NDArray[Any]
+                if isinstance(pd_array, pd.Categorical):
+                    array = pd_array
+                else:
+                    array = convert_pd_array_to_np_array(pd_array)
+                r_series = self.convert_to_r_object(array)
+                r_value.append(r_series)
+
+            index = data.index
+            if isinstance(index, pd.RangeIndex):
+                assert isinstance(index.start, int)
+                if (index.start == 1
+                    and index.stop == data.shape[0] + 1
+                    and index.step == 1
+                ):
+                    row_names = np.ma.array(
+                            data=[R_INT_NA, -data.shape[0]],
+                            mask=[True, False],
+                            fill_value=R_INT_NA,
+                        )
+                else:
+                    row_names = range(index.start, index.stop, index.step)
+            elif isinstance(index, pd.Index):
+                if (index.dtype == "object"
+                    or np.issubdtype(str(index.dtype), np.integer)):
+                    row_names = index.to_numpy()
+                else:
+                    msg = f"pd.DataFrame pd.Index {index.dtype} not implemented"
+                    raise NotImplementedError(msg)
+            else:
+                msg = f"pd.DataFrame index {type(index)} not implemented"
+                raise NotImplementedError(msg)
+
+            attributes = {
+                "names": np.array(column_names, dtype=np.dtype("U")),
+                "class": "data.frame",
+                "row.names": row_names,
+            }
+            if self.df_attr_order is not None:
+                attributes = {k: attributes[k] for k in self.df_attr_order}
+
+        else:
+            msg = f"type {type(data)} not implemented"
+            raise NotImplementedError(msg)
 
-    minimum_version_with_encoding = 3
-    extra = (RExtraInfo(encoding.upper())
-             if versions.format >= minimum_version_with_encoding
-             else RExtraInfo(None))
+        if attributes is not None:
+            r_attributes = self.convert_to_r_attributes(attributes)
+        else:
+            r_attributes = None
 
-    return RData(versions, extra, r_object)
+        return build_r_object(r_type, value=r_value,
+                              is_object=is_object,
+                              attributes=r_attributes,
+                              tag=tag, gp=gp)
```

Comment:
I have seen this at some points in this PR, but in this case is particularly noticeable. In general, for multiline lines I prefer to have the ending parenthesis (or bracket or brace) in its own line, with the same indentation as the line in which it is opened. Also, it is better for each parameter to be in its own line, for improved readability:
```suggestion
        return build_r_object(
            r_type,
            value=r_value,
            is_object=is_object,
            attributes=r_attributes,
            tag=tag,
            gp=gp,
        )
```

Reply:
I agree. This reads much better.